google/api_core/retry_streaming.py

```diff
@@ -0,0 +1,183 @@
+            on_error: A function to call while processing a
+                retryable exception.  Any error raised by this function will *not*
+                be caught.
+        """
+        self.target_fn = target
+        self.active_target: Iterable = self.target_fn()
+        self.predicate = predicate
+        self.sleep_generator = iter(sleep_generator)
+        self.on_error = on_error
+        self.timeout = timeout
+        if self.timeout is not None:
+            self.deadline = datetime_helpers.utcnow() + datetime.timedelta(
+                seconds=self.timeout
+            )
+        else:
+            self.deadline = None
+
+    def __iter__(self):
+        """
+        Implement the iterator protocol.
+        """
+        return self
+
+    def _handle_exception(self, exc):
+        """
+        When an exception is raised while iterating over the active_target,
+        check if it is retryable. If so, create a new active_target and
+        continue iterating. If not, raise the exception.
+        """
+        if not self.predicate(exc):
+            raise exc
+        else:
+            # run on_error callback if provided
+            if self.on_error:
+                self.on_error(exc)
+            try:
+                next_sleep = next(self.sleep_generator)
+            except StopIteration:
+                raise ValueError("Sleep generator stopped yielding sleep values")
+            # if deadline is exceeded, raise RetryError
+            if self.deadline is not None:
+                next_attempt = datetime_helpers.utcnow() + datetime.timedelta(
+                    seconds=next_sleep
+                )
+                if self.deadline < next_attempt:
+                    raise exceptions.RetryError(
+                        f"Deadline of {self.timeout:.1f} seconds exceeded", exc
+                    ) from exc
+            # sleep before retrying
+            _LOGGER.debug(
+                "Retrying due to {}, sleeping {:.1f}s ...".format(exc, next_sleep)
+            )
+            time.sleep(next_sleep)
+            self.active_target = self.target_fn()
+
+    def __next__(self):
+        """
+        Implement the iterator protocol.
+        """
+        try:
+            return next(self.active_target)
+        except Exception as exc:
+            self._handle_exception(exc)
+        # if retryable exception was handled, try again with new active_target
+        return self.__next__()
+
+    def close(self):
+        """
+        Close the active_target if supported. (e.g. target is a generator)
+
+        Raises:
+          - AttributeError if the active_target does not have a close() method
+        """
+        if getattr(self.active_target, "close", None):
+            return self.active_target.close()
+        else:
+            raise AttributeError(
+                "close() not implemented for {}".format(self.active_target)
+            )
+
+    def send(self, value):
```

Comment:
Add type hint for return value. Add information in the docstring to describe the arg `value`.